pyanaconda/controller.py

```diff
@@ -0,0 +1,124 @@
+from pyanaconda.iutil import synchronized
+import logging
+log = logging.getLogger("anaconda")
+
+controllers = {}
+controller_categories_map = {}
+
+class Controller(object):
+    """A singleton that track initialization of Anaconda modules."""
+    def __init__(self):
+        self._lock = RLock()
+        self._modules = set()
+        self._all_modules_added = False
+        self.init_done = Signal()
+        self._init_done_triggered = False
+        self._added_module_count = 0
+
+    @synchronized
+    def module_init_start(self, module):
+        """Tell the controller that a module has started initialization.
+
+        :param module: a module which has started initialization
+        """
+        if self._all_modules_added:
+            log.warning("Late module_init_start() from: %s", self)
+        elif module in self._modules:
+            log.warning("Module already marked as initializing: %s", module)
+        else:
+            self._added_module_count += 1
+            self._modules.add(module)
+
+    @synchronized
+    def all_modules_added(self):
+        """Tell the controller that all expected modules have started initialization.
+
+        Tell the controller that all expected modules have been registered
+        for initialization tracking (or have already been initialized)
+        and no more are expected to be added.
+
+        This is needed so that we don't prematurely trigger the init_done signal
+        when all known modules finish initialization while other modules have not
+        yet been added.
+        """
+        log.info("Initialization of all modules (%d) has been started.", self._added_module_count)
+        self._all_modules_added = True
+
+        # if all modules finished initialization before this was added then
+        # trigger the init_done signal at once
+        if not self._modules and not self._init_done_triggered:
+            self._trigger_init_done()
+
+    def module_init_done(self, module):
+        """Tell the controller that a module has finished initialization.
+
+        An if no more modules are being initialized trigger the init_done signal.
+
+        :param module: a module that has finished initialization
+        """
+        init_done = False
+        with self._lock:
+            # prevent the init_done signal from
+            # being triggered more than once
+            if self._init_done_triggered:
+                log.warning("Late module_init_done from module %s.", module)
+                return
+            if module in self._modules:
+                log.info("Module initialized: %s", module)
+                self._modules.discard(module)
+            else:
+                log.warning("Unknown module reported as initialized: %s", module)
+            # don't trigger the signal if all modules have not yet been added
+            if self._all_modules_added and not self._modules:
+                init_done = True
+                self._init_done_triggered = True
```

Comment:
It should be done inside of the self._lock mutex as I *think* there could be an issue if the signal was triggered while the lock is held. 

That's why the `_trigger_init_done()` method, which does the signal triggering, is called outside of the lock. 

But in any case, the code could be written better & better documented so I'll try to improve it.

Also `all_modules_added()` might call `_trigger_init_done()` as well, so I'll make sure to also set `self._init_done_triggered` to True there.